aem-authoring-toolkit-plugin/src/main/java/com/exadel/aem/toolkit/core/handlers/widget/AccordionWidgetHandler.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.exadel.aem.toolkit.core.handlers.widget;
+
+import com.exadel.aem.toolkit.api.annotations.container.AccordionPanel;
+import com.exadel.aem.toolkit.api.annotations.container.PlaceOn;
+import com.exadel.aem.toolkit.api.annotations.main.JcrConstants;
+import com.exadel.aem.toolkit.api.annotations.meta.ResourceTypes;
+import com.exadel.aem.toolkit.api.annotations.widgets.AccordionWidget;
+import com.exadel.aem.toolkit.core.exceptions.InvalidSettingException;
+import com.exadel.aem.toolkit.core.handlers.Handler;
+import com.exadel.aem.toolkit.core.maven.PluginRuntime;
+import com.exadel.aem.toolkit.core.util.DialogConstants;
+import com.exadel.aem.toolkit.core.util.PluginObjectPredicates;
+import com.exadel.aem.toolkit.core.util.PluginObjectUtility;
+import com.exadel.aem.toolkit.core.util.PluginXmlContainerUtility;
+import com.google.common.collect.ImmutableMap;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.jcr.resource.api.JcrResourceConstants;
+import org.w3c.dom.Element;
+
+import java.lang.reflect.Field;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * {@link Handler} implementation used to create markup responsible for Granite {@code Accordion} widget functionality
+ * within the {@code cq:dialog} XML node
+ */
+class AccordionWidgetHandler implements WidgetSetHandler {
+    private static final String DEFAULT_TAB_NAME = "accordion";
+    private static final String EMPTY_ACCORDION_EXCEPTION_MESSAGE = "No valid fields found in accordion class ";
+    private static final String NO_TABS_DEFINED_EXCEPTION_MESSAGE = "No accordions defined for the dialog at ";
+
+    /**
+     * Processes the user-defined data and writes it to XML entity
+     *
+     * @param element Current XML element
+     * @param field   Current {@code Field} instance
+     */
+    @Override
+    public void accept(Element element, Field field) {
```

Comment:
I haven't completely grasped the code as yet, but this seems very much (not 100% but 80% identical) to the method in TabsWidgetHandler, and both are much similar to `ContainerHandler.acceptParent()`. In particular, the part starting from "getContainerFields()".

It is not merely the matter of cleanliness (however partially too). It's the matter of refactoring all three files if something goes wrong.

I myself have many times had this head-banging situation like "what again? I bet I've already changed this crap! Oh, my, it was another place..."